odl/discr/discr_utils.py

```diff
@@ -0,0 +1,812 @@
+    def linear_interp(arg, out=None):
+        """Interpolating function with vectorization."""
+        if is_valid_input_meshgrid(arg, f.ndim):
+            input_type = 'meshgrid'
+        else:
+            input_type = 'array'
+
+        interpolator = _LinearInterpolator(
+            coord_vecs,
+            f,
+            input_type=input_type
+        )
+
+        return interpolator(arg, out=out)
+
+    return linear_interp
+
+
+def per_axis_interpolator(f, coord_vecs, schemes, nn_variants=None):
+    """Return a per axis defined interpolator for discrete values.
+    With this function, the interpolation scheme can be chosen for each axis
+    separately.
+    Parameters
+    ----------
+    f : numpy.ndarray
+        Function values that should be interpolated.
+    coord_vecs : sequence of numpy.ndarray
+        Coordinate vectors of the rectangular grid on which interpolation
+        should be based. They must be sorted in ascending order. Usually
+        they are obtained as ``grid.coord_vectors`` from a `RectGrid`.
+    schemes : string or sequence of strings
+        Indicates which interpolation scheme to use for which axis.
+        A single string is interpreted as a global scheme for all
+        axes.
+    nn_variants : string or sequence of strings, optional
+        Which variant ('left' or 'right') to use in nearest neighbor
+        interpolation for which axis. A single string is interpreted
+        as a global variant for all axes.
+        This option has no effect for schemes other than nearest
+        neighbor.
+    Examples
+    --------
+    Choose linear interpolation in the first axis and nearest neighbor in
+    the second:
+    >>> part = odl.uniform_partition([0, 0], [1, 5], shape=(2, 4))
+    >>> part.coord_vectors
+    (array([ 0.25,  0.75]), array([ 0.625,  1.875,  3.125,  4.375]))
+    >>> f = np.array([[1, 2, 3, 4],
+    ...               [5, 6, 7, 8]],
+    ...              dtype=float)
+    >>> interpolator = per_axis_interpolator(
+    ...     f, part.coord_vectors, ['linear', 'nearest']
+    ... )
+    >>> interpolator([1, 1])  # single point
+    2.5
+    >>> interpolator([[0.5, 2.0],
+    ...               [0.0, 4.5],
+    ...               [0.0, 3.0]])  # 3 points at once
+    array([ 6. , -7.5,  1.5])
+    >>> from odl.discr.grid import sparse_meshgrid
+    >>> mesh = sparse_meshgrid([0.0, 0.5, 1.0], [1.5, 3.5])
+    >>> interpolator(mesh)  # 3x2 grid of points
+    array([[ 1. ,  1.5],
+           [ 4. ,  5. ],
+           [ 3. ,  3.5]])
+    """
+    f = np.asarray(f)
+
+    schemes_in = schemes
+    if is_string(schemes):
+        scheme = str(schemes).lower()
+        if scheme not in _SUPPORTED_INTERP_SCHEMES:
+            raise ValueError('`schemes` {!r} not understood'
+                             ''.format(schemes_in))
+        schemes = [scheme] * f.ndim
+    else:
+        schemes = [str(scm).lower() if scm is not None else None
```

Comment:
~~Check that they are all in `_SUPPORTED_INTERP_SCHEMES`?~~ Seems to be done in line 431. Make that clear here?